src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+AsyncTransferFacilitator::~AsyncTransferFacilitator()
+{
+    Reset();
+}
+
+void AsyncTransferFacilitator::Reset()
+{
+    mExchange.Release();
+    mTransfer.Reset();
+}
+
+CHIP_ERROR AsyncTransferFacilitator::SendMessage(TransferSession::OutputEvent & event)
+{
+
+    VerifyOrReturnError(mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    Messaging::SendFlags sendFlags;
+
+    // All messages sent from the Sender expect a response, except for a StatusReport which would indicate an error and
+    // the end of the transfer.
+    if (!event.msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kExpectResponse);
+    }
+
+    auto & msgTypeData = event.msgTypeData;
+
+    Messaging::ExchangeContext * ec = mExchange.Get();
+    VerifyOrReturnError(ec != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    // Set the response timeout on the exchange before sending the message.
+    ec->SetResponseTimeout(mTimeout);
+
+    CHIP_ERROR err = ec->SendMessage(msgTypeData.ProtocolId, msgTypeData.MessageType, std::move(event.MsgData), sendFlags);
+
+    // If there's an error sending the message, release the exchange.
```

Comment:
The comment should be documenting _why_ we are doing this, not the obvious what-we-are-doing.

But also: I am very much confused about the error-handling here.  If this fails, what does the cleanup?  That should be documented here.

Reply:
We should clean up if this fails. I will add that.